File: core/src/main/java/fr/sncf/osrd/api/pathfinding/RemainingDistanceEstimator.java
package fr.sncf.osrd.api.pathfinding;

import fr.sncf.osrd.geom.Point;
import fr.sncf.osrd.infra.api.signaling.SignalingRoute;
import fr.sncf.osrd.utils.graph.functional_interfaces.AStarHeuristic;
import fr.sncf.osrd.utils.graph.Pathfinding;
import java.util.ArrayList;
import java.util.Collection;

/** This is a function object that estimates the remaining distance to the closest target point,
 * using geo data. It is used as heuristic for A*. */
public class RemainingDistanceEstimator implements AStarHeuristic<SignalingRoute> {

    private final Collection<Point> targets;
    private final double remainingDistance;

    /** Targets closer than this threshold will be merged together */
    private static final double distanceThreshold = 1.;

    /** Constructor */
    public RemainingDistanceEstimator(
            Collection<Pathfinding.EdgeLocation<SignalingRoute>> edgeLocations,
            double remainingDistance
    ) {
        targets = new ArrayList<>();
        for (var edgeLocation : edgeLocations) {
            var point = routeOffsetToPoint(edgeLocation.edge(), edgeLocation.offset());
            var skip = false;
            for (var otherPoint : targets) {
                if (point.distanceAsMeters(otherPoint) < distanceThreshold) {
                    skip = true; // Avoid adding duplicate geo points to the target list
                    break;
                }
            }
            if (!skip)
                targets.add(point);
        }
        this.remainingDistance = remainingDistance;
    }

    /** Converts a route and offset from its start into a geo point */
    private static Point routeOffsetToPoint(SignalingRoute route, double pointOffset) {
        for (var trackRange : route.getInfraRoute().getTrackRanges()) {
            if (pointOffset <= trackRange.getLength()) {
                var trackLocation = trackRange.offsetLocation(pointOffset);
                var normalizedOffset = trackLocation.offset() / trackLocation.track().getLength();
                var geo = trackLocation.track().getGeo();
                if (geo == null)
                    return new Point(0, 0);
                return geo.interpolateNormalized(normalizedOffset);
            }
            pointOffset -= trackRange.getLength();
        }
        throw new RuntimeException("Couldn't find offset on route");
    }

    /** Compute the minimum geo distance between two steps */
    public static double minDistanceBetweenSteps(
            Collection<Pathfinding.EdgeLocation<SignalingRoute>> step1,
            Collection<Pathfinding.EdgeLocation<SignalingRoute>> step2
    ) {
        var step1Points = step1.stream()
                .map(loc -> routeOffsetToPoint(loc.edge(), loc.offset()))
                .toList();
        var step2Points = step2.stream()
                .map(loc -> routeOffsetToPoint(loc.edge(), loc.offset()))
                .toList();

        var res = Double.POSITIVE_INFINITY;
        for (var point1: step1Points) {
            for (var point2: step2Points) {
                res = Double.min(res, point1.distanceAsMeters(point2));
            }
        }
        return res;
    }

    @Override
    public double apply(SignalingRoute signalingRoute, double offset) {
        var res = Double.POSITIVE_INFINITY;
        var point = routeOffsetToPoint(signalingRoute, offset);
        for (var target : targets)
            res = Double.min(res, point.distanceAsMeters(target));
        return res + remainingDistance;
    }
}

File: core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/RemainingDistanceEstimator.kt
package fr.sncf.osrd.api.pathfinding

import fr.sncf.osrd.geom.Point
import fr.sncf.osrd.graph.PathfindingEdgeLocationId
import fr.sncf.osrd.sim_infra.api.Block
import fr.sncf.osrd.sim_infra.api.BlockId
import fr.sncf.osrd.sim_infra.api.BlockInfra
import fr.sncf.osrd.sim_infra.api.RawSignalingInfra
import fr.sncf.osrd.utils.Direction
import fr.sncf.osrd.utils.units.Distance
import fr.sncf.osrd.utils.units.Offset
import fr.sncf.osrd.utils.units.meters
import kotlin.math.min

/**
 * This is a function object that estimates the remaining distance to the closest target point,
 * using geo data. It is used as heuristic for A*.
 */
class RemainingDistanceEstimator(
    private val blockInfra: BlockInfra,
    private val rawInfra: RawSignalingInfra,
    edgeLocations: Collection<PathfindingEdgeLocationId<Block>>,
    private val remainingDistance: Distance
) {
    private val targets: MutableCollection<Point> = ArrayList()

    init {
        for (edgeLocation in edgeLocations) {
            val point =
                blockOffsetToPoint(blockInfra, rawInfra, edgeLocation.edge, edgeLocation.offset)
            // Avoid adding duplicate geo points to the target list
            val isDuplicate = targets.any { point.distanceAsMeters(it) < DISTANCE_THRESHOLD }
            if (!isDuplicate) targets.add(point)
        }
    }

    fun apply(edge: BlockId, offset: Offset<Block>): Distance {
        var resMeters = Double.POSITIVE_INFINITY
        val point = blockOffsetToPoint(blockInfra, rawInfra, edge, offset)
        for (target in targets) resMeters = min(resMeters, point.distanceAsMeters(target))

        return resMeters.meters + remainingDistance
    }

    companion object {
        /** Targets closer than this threshold will be merged together */
        private const val DISTANCE_THRESHOLD = 1.0
    }
}

/**
 * Compute the minimum geo distance between two steps. Expected to be used when instantiating the
 * heuristic, to estimate the remaining total distance for any step.
 */
fun minDistanceBetweenSteps(
    blockInfra: BlockInfra,
    rawInfra: RawSignalingInfra,
    step1: Collection<PathfindingEdgeLocationId<Block>>,
    step2: Collection<PathfindingEdgeLocationId<Block>>
): Distance {
    val step1Points = step1.map { blockOffsetToPoint(blockInfra, rawInfra, it.edge, it.offset) }
    val step2Points = step2.map { blockOffsetToPoint(blockInfra, rawInfra, it.edge, it.offset) }
    var res = Double.POSITIVE_INFINITY
    for (point1 in step1Points) {
        for (point2 in step2Points) {
            res = min(res, point1.distanceAsMeters(point2))
        }
    }

    return res.meters
}

/** Converts a block and offset from its start into a geo point */
private fun blockOffsetToPoint(
    blockInfra: BlockInfra,
    rawInfra: RawSignalingInfra,
    blockIdx: BlockId,
    pointOffset: Offset<Block>
): Point {
    // Ideally, we would just call `makePathProps(blockId).getGeo()` to get geo data for the block.
    // But that introduces a lot of overhead (building the path and projecting the results),
    // which results in a net performance loss when using the heuristic.
    // This implementation is more verbose but quite lightweight.
    // TODO: maybe investigate where the path overhead is coming from, it may not be normal

    var pathOffset = 0.meters
    for (chunk in blockInfra.getTrackChunksFromBlock(blockIdx)) {
        val chunkLength = rawInfra.getTrackChunkLength(chunk.value)
        if (pathOffset + chunkLength.distance < pointOffset.distance) {
            pathOffset += chunkLength.distance
            continue
        }

        val remainingOffsetOnBlock = pointOffset.distance - pathOffset
        val chunkOffset =
            if (chunk.direction == Direction.INCREASING) {
                remainingOffsetOnBlock
            } else {
                chunkLength.distance - remainingOffsetOnBlock
            }
        val lineString = rawInfra.getTrackChunkGeom(chunk.value)
        return lineString.interpolateNormalized(chunkOffset.meters / chunkLength.distance.meters)
    }
    throw RuntimeException("Unreachable (block offset outside of block)")
}
